research/deep_speech/deep_speech.py

```diff
@@ -103,45 +137,76 @@ def evaluate_model(
   return eval_results
 
 
-def convert_keras_to_estimator(keras_model, num_gpus):
-  """Configure and convert keras model to Estimator.
+def model_fn(features, labels, mode, params):
+  """Define model function for deep speech model.
 
   Args:
-    keras_model: A Keras model object.
-    num_gpus: An integer, the number of GPUs.
+    features: a dictionary of input_data features. It includes the data
+      input_length, label_length and the spectrogram features.
+    labels: a list of labels for the input data.
+    mode: current estimator mode; should be one of
+      `tf.estimator.ModeKeys.TRAIN`, `EVALUATE`, `PREDICT`.
+    params: a dict of hyper parameters to be passed to model_fn.
 
   Returns:
-    estimator: The converted Estimator.
+    EstimatorSpec parameterized according to the input params and the
+    current mode.
   """
-  # keras optimizer is not compatible with distribution strategy.
-  # Use tf optimizer instead
-  optimizer = tf.train.MomentumOptimizer(
-      learning_rate=flags_obj.learning_rate, momentum=flags_obj.momentum,
-      use_nesterov=True)
-
-  # ctc_loss is wrapped as a Lambda layer in the model.
-  keras_model.compile(
-      optimizer=optimizer, loss={"ctc_loss": lambda y_true, y_pred: y_pred})
-
-  distribution_strategy = distribution_utils.get_distribution_strategy(
-      num_gpus)
-  run_config = tf.estimator.RunConfig(
-      train_distribute=distribution_strategy)
-
-  estimator = tf.keras.estimator.model_to_estimator(
-      keras_model=keras_model, model_dir=flags_obj.model_dir, config=run_config)
-
-  return estimator
+  num_classes = params["num_classes"]
+  input_length = features["input_length"]
+  label_length = features["label_length"]
+  features = features["features"]
+
+  # Create DeepSpeech2 model.
+  model = deep_speech_model.DeepSpeech2(
+      flags_obj.rnn_hidden_layers, flags_obj.rnn_type,
+      flags_obj.is_bidirectional, flags_obj.rnn_hidden_size,
+      num_classes, flags_obj.use_bias)
+
+  if mode == tf.estimator.ModeKeys.PREDICT:
+    logits = model(features, training=False)
+    predictions = {
+        "classes": tf.argmax(logits, axis=2),
+        "probabilities": tf.nn.softmax(logits),
+        "logits": logits
+    }
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        predictions=predictions)
+
+  # In training mode.
```

Comment:
I guess you don't handle the EVAL model here, which probably should throw exception if the mode == EVALUATE.

Reply:
Raise error here instead.

Reply:
The EVAL mode will be handled by the following codes if someone calls estimator.evaluate(), and it just has no eval metrics (which is not mandatory in eval mode). I don't think raise error is a good idea.